### PR TITLE
pc - add bootstrap from cdn plus tests

### PR DESCRIPTION
**Pull request by pconrad**
This pull requests adds in Bootstrap from a CDN, plus some tests to ensure that the web pages load correctly, and have the CSS and JavaScript files needed for Bootstrap.

These tests cases provide us with a safety net for the next step, which is try to refactor the template files to dry up the inclusion of the bootstrap elements, so that we don't need to repeat them in every page.

### CODE REVIEW

File: src/test/java/hello/HomePageTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Test
+    public void getHomePage_IsOk() throws Exception {
+        mvc.perform(MockMvcRequestBuilders.get("/").accept(MediaType.TEXT_HTML))
+                .andExpect(status().isOk())
+                .andExpect(content().contentType("text/html;charset=UTF-8"));
+    }
+
+    @Test
+    public void getHomePage_ContentType() throws Exception {
```
Comment (scottpchow23) [inline src/test/java/hello/HomePageTest.java:28]: Duplicate of the above test; I'd personally bundle `IsOk` and `ContentType` given that if either of these fail it's likely to be either a catastrophic change or a red herring.

File: src/test/java/hello/GreetingControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    public void getGreeting_isOk() throws Exception {
+        mvc.perform(MockMvcRequestBuilders.get("/greeting").accept(MediaType.TEXT_HTML))
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    public void getGreeting_ContentType() throws Exception {
+        mvc.perform(MockMvcRequestBuilders.get("/greeting").accept(MediaType.TEXT_HTML))
+                .andExpect(status().isOk())
```
Comment (scottpchow23) [inline src/test/java/hello/GreetingControllerTest.java:30]: The same comment as below, even if this test and one above are slightly different. Checking `status().isOk()` for the same request isn't necessary.

File: src/test/java/hello/HomePageTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            String jsXPath = String.format("//script[@src='%s']",s);
+            mvc.perform(MockMvcRequestBuilders.get("/greeting").accept(MediaType.TEXT_HTML))
+              .andExpect(status().isOk())
+              .andExpect(xpath(jsXPath).exists());
+        }
+    }
+
+
+    @Test
+    public void getHomePage_hasCorrectTitle() throws Exception {
```
Comment (scottpchow23) [inline src/test/java/hello/HomePageTest.java:50]: It seems that these tests are boilerplate that has to go on every single page that we include javascript and bootstrap. That being said, I'm not away of any way to dry this up in JUnit given that it relies on one function = one test.

This comment is more observation than review.

**Review by scottpchow23 — CHANGES_REQUESTED** [review]

File: src/test/java/hello/HomePageTest.java (hunk 1 above)
Comment (pconrad) [reply src/test/java/hello/HomePageTest.java:28]: Agreed..  there is a lot of drying up that I plan to do in the next iteration.

File: src/test/java/hello/GreetingControllerTest.java (hunk 2 above)
Comment (pconrad) [reply src/test/java/hello/GreetingControllerTest.java:30]: My thought here was that before I start testing specific content, I want to be sure that the request is sane.     In this case, I'd argue that the isOk is reasonable.   My philosophy is that tests don't need to be perfectly dry---they need to be optimized to give useful feedback to humans about what's wrong.

So some degree of redundant testing is ok in my view.   There is a spectrum here, though and there certainly is some amount of drying up that can be done.

File: src/test/java/hello/HomePageTest.java (hunk 3 above)
Comment (pconrad) [reply src/test/java/hello/HomePageTest.java:50]: My intent in the next iteration is to factor out this code into something included on every page.  I'm not expecting that we'd have these tests for every page as the app grows, but I wanted some way to be sure that as I refactor, I don't screw things up and have the Boostrap loaded on one page but not another.

We might be able to refactor this into a simple "assertBootStrapIsLoaded(page)" call once we understand better how the testing methods work in Spring.

Thanks for the feedback

File: src/test/java/hello/GreetingControllerTest.java (hunk 2 above)
Comment (pconrad) [reply src/test/java/hello/GreetingControllerTest.java:30]: Actually, I do think it's reasonable to combine the tests for isOk and contentType and just have the contentType test. 

My comment above was more about the "isOk" at line 43 in the test for "isBootstrapLoaded".    The idea is to test whether the page returned at all before looking for specific content.

File: src/test/java/hello/HomePageTest.java (hunk 1 above)
Comment (pconrad) [reply src/test/java/hello/HomePageTest.java:28]: I did remove the separate isOk tests (as separate tests) from both test files.

**Pull request merged**